File: Ai algorithm implementation/Genetic_Algorithm_(N-Queens).cpp
#include <bits/stdc++.h>
using namespace std;

int N = 8;
int POP_SIZE = 100;
int GENERATIONS = 1000;
double MUTATION_RATE = 0.1;

int heuristic(vector<int> &state) {
    int h = 0;
    for(int i=0;i<N;i++)
        for(int j=i+1;j<N;j++)
            if(state[i]==state[j] || abs(state[i]-state[j])==abs(i-j)) h++;
    return h;
}

// Crossover: single-point
pair<vector<int>, vector<int>> crossover(vector<int> &p1, vector<int> &p2) {
    int point = rand()%N;
    vector<int> c1 = p1, c2 = p2;
    for(int i=point;i<N;i++){
        swap(c1[i], c2[i]);
    }
    return {c1,c2};
}

// Mutation: random column change
void mutate(vector<int> &child) {
    if((double)rand()/RAND_MAX < MUTATION_RATE){
        int col = rand()%N;
        child[col] = rand()%N;
    }
}

int main() {
    srand(time(0));
    // Initialize population
    vector<vector<int>> population(POP_SIZE, vector<int>(N));
    for(auto &ind: population)
        for(int i=0;i<N;i++) ind[i]=rand()%N;

    for(int gen=0; gen<GENERATIONS; gen++){
        // Sort by fitness (lower heuristic better)
        sort(population.begin(), population.end(), [](vector<int> &a, vector<int> &b){
            return heuristic(a) < heuristic(b);
        });

        if(heuristic(population[0])==0){
            cout<<"Solution found at generation "<<gen<<"\n";
            break;
        }

        vector<vector<int>> newPop;
        // Elitism: keep top 10%
        int elite = POP_SIZE/10;
        for(int i=0;i<elite;i++) newPop.push_back(population[i]);

        // Generate rest of new population
        while(newPop.size()<POP_SIZE){
            int i1 = rand()%elite, i2 = rand()%elite;
            auto children = crossover(population[i1], population[i2]);
            mutate(children.first);
            mutate(children.second);
            newPop.push_back(children.first);
            if(newPop.size()<POP_SIZE) newPop.push_back(children.second);
        }

        population = newPop;
    }

    cout << "Genetic Algorithm Solution (Heuristic=" << heuristic(population[0]) << "):\n";
    for(int i=0;i<N;i++) cout << population[0][i] << " ";
    cout << "\n";
}

File: Ai algorithm implementation/Genetic_Algorithm_(N-Queens).py
import random

N = 8
POP_SIZE = 100
GENERATIONS = 1000
MUTATION_RATE = 0.1

def heuristic(state):
    h = 0
    for i in range(N):
        for j in range(i + 1, N):
            if state[i] == state[j] or abs(state[i] - state[j]) == abs(i - j):
                h += 1
    return h

def crossover(p1, p2):
    point = random.randint(0, N - 1)
    c1 = p1[:point] + p2[point:]
    c2 = p2[:point] + p1[point:]
    return c1, c2

def mutate(child):
    if random.random() < MUTATION_RATE:
        col = random.randint(0, N - 1)
        child[col] = random.randint(0, N - 1)

random.seed()
population = [[random.randint(0, N - 1) for _ in range(N)] for _ in range(POP_SIZE)]

for gen in range(GENERATIONS):
    population.sort(key=lambda ind: heuristic(ind))
    
    if heuristic(population[0]) == 0:
        print(f"Solution found at generation {gen}")
        break
    
    new_pop = []
    elite = POP_SIZE // 10
    new_pop.extend(population[:elite])
    
    while len(new_pop) < POP_SIZE:
        i1 = random.randint(0, elite - 1)
        i2 = random.randint(0, elite - 1)
        c1, c2 = crossover(population[i1], population[i2])
        mutate(c1)
        mutate(c2)
        new_pop.append(c1)
        if len(new_pop) < POP_SIZE:
            new_pop.append(c2)
    
    population = new_pop

print(f"Genetic Algorithm Solution (Heuristic={heuristic(population[0])}):")
print(population[0])
